### PR TITLE
`sample_probs` is extracted out of `sample_state`

### PR DESCRIPTION
**Pull request by JerryChen97**
**Context:**
Currently, the sample_state helper function accepts a state vector, turns it into probabilities, and then samples it. This task is to break out the code for sample_probs from this function.

By breaking this code into it's own helper function, we can reuse it for a "sample density matrix" implementation.
**Description of the Change:**
 - [x] Separate `devices.qubit.sampling.sample_probs` from `sample_state`
 - [x] Separate `devices.qubit.sampling.sample_probs_jax` from `_sample_state_jax`
 - [x] similar but for qutrit mixed
 - [x] similar but for qutrit mixed

**Benefits:**
Better modularization; disentanglement

Future usage in new devices

**Possible Drawbacks:**

**Related GitHub Issues:**

**Related ShortCut Stories:**
[sc-73317]

### CODE REVIEW

**Comment by JerryChen97** [comment]
Reading the previous [testcases](https://github.com/PennyLaneAI/pennylane/blob/c36d999d5c9d149ccb190305851b264af3b14055/tests/devices/qubit/test_sampling.py#L733C1-L761C40) regarding `batched_state_not_normalized` I realized that it seems that we are simply rejecting any states with `norm_err>cutoff =1e-07`? If so can I alter the norm check [logic](https://github.com/PennyLaneAI/pennylane/blob/c36d999d5c9d149ccb190305851b264af3b14055/pennylane/devices/qubit/sampling.py#L523C1-L541C29) such that we first check this thing and if not satisfied immediately raise error? @albi3ro

**Comment by JerryChen97** [comment]
> Reading the previous [testcases](https://github.com/PennyLaneAI/pennylane/blob/c36d999d5c9d149ccb190305851b264af3b14055/tests/devices/qubit/test_sampling.py#L733C1-L761C40) regarding `batched_state_not_normalized` I realized that it seems that we are simply rejecting any states with `norm_err > cutoff = 1e-07`? If so can I alter the norm check [logic](https://github.com/PennyLaneAI/pennylane/blob/c36d999d5c9d149ccb190305851b264af3b14055/pennylane/devices/qubit/sampling.py#L523C1-L541C29) such that we first check this thing and if not satisfied immediately raise error? @albi3ro

Also I saw some tests expect ValueError but deliberately avoid 'jax' beacuse jax choice accepts unnormalized probs. This very confusing, why don't we just preprocess this to raise corresponding error?

**Comment by JerryChen97** [comment]
Short summary:
1. Right now it's just naive decoupling basically move the processing for probs to outside
2. One modification was made such that the norm error sanity check is preprocessed instead of tangled inside the main work
3. Personally I would recommend further decomposition since the process of sampling over a well-defined spin system with probs already given should be universal and wrapped into somewhere in math module. But that would go beyond current scope; need another story.

File: pennylane/devices/qubit/sampling.py
Comment (albi3ro) [inline pennylane/devices/qubit/sampling.py:1]: Since the "state to probs" part of `sample_state` and `sample_state_jax` are basically the exact same, I'm wondering if we should move the dispatch between the jax-based sampling and the numpy-based sampling *inside* of `sample_probs`. 

So we would have one:
```
sample_probs(probs, shots, num_wires, is_state_batched, rng, prng_key)
```
that dispatches between a jax based implementation and a numpy based implementation based on the interface of `probs` and the existence of a `prng_key`. Then `_sample_probs_jax` would just be a private helper.

What would you think?

**Review by albi3ro — COMMENTED** [review]
I also think we should make `sample_probs` accessible in `qml.devices.qubit.sample_probs`.  This will make it a bit more visible and accessible.

Let's add it as an import and to the docs table of contents in `pennylane/devices/qubit/__init__.py`.

**Comment by JerryChen97** [comment]
> I also think we should make `sample_probs` accessible in `qml.devices.qubit.sample_probs`. This will make it a bit more visible and accessible.
> 
> Let's add it as an import and to the docs table of contents in `pennylane/devices/qubit/__init__.py`.

Can't agree more! Will do it

File: pennylane/devices/qubit/__init__.py [hunk 1]
```diff
@@ -36,5 +36,5 @@
 from .adjoint_jacobian import adjoint_jacobian, adjoint_jvp, adjoint_vjp
 from .initialize_state import create_initial_state
 from .measure import measure
-from .sampling import sample_state, measure_with_samples
+from .sampling import sample_state, measure_with_samples, sample_probs
```
Comment (albi3ro) [inline pennylane/devices/qubit/__init__.py:39]: And add to the `toctree` above too

File: pennylane/devices/qubit/sampling.py [hunk 2]
```diff
@@ -547,36 +547,22 @@ def _sample_state_jax(
...
-    basis_states = np.arange(2**num_wires)
-
-    flat_state = flatten_state(state, total_indices)
-    with qml.queuing.QueuingManager.stop_recording():
-        probs = qml.probs(wires=wires_to_sample).process_state(flat_state, state_wires)
+    basis_states = jnp.arange(2**num_wires)
 
     if is_state_batched:
-        keys = jax_random_split(prng_key, num=len(state))
+        keys = jax.random.split(prng_key, num=probs.shape[0])
```
Comment (mudit2812) [inline pennylane/devices/qubit/sampling.py:557]: If the `jax_random_split` function is not being used anymore, we can remove it. If it is used elsewhere, then I think it's best to continue using it instead of `jax.random.split` for consistency.

File: pennylane/devices/qubit/sampling.py [hunk 3]
```diff
@@ -547,36 +547,22 @@ def _sample_state_jax(
...
         samples = jnp.array(
             [
                 jax.random.choice(_key, basis_states, shape=(shots,), p=prob)
                 for _key, prob in zip(keys, probs)
             ]
         )
     else:
-        _, key = jax_random_split(prng_key)
-        samples = jax.random.choice(key, basis_states, shape=(shots,), p=probs)
+        samples = jax.random.choice(prng_key, basis_states, shape=(shots,), p=probs)
```
Comment (mudit2812) [inline pennylane/devices/qubit/sampling.py:565]: We should continue splitting the PRNGKey here, because otherwise the samples between each execution will not differ, which we need for MCMs to work correctly.

File: pennylane/devices/qubit/sampling.py [hunk 4]
```diff
@@ -471,67 +479,60 @@ def sample_state(
...
-                normalize_condition = True
-            if s > cutoff:
-                normalize_condition = False
-                break
+def sample_probs(probs, shots, num_wires, is_state_batched, rng, prng_key=None):
+    """
+    Sample from given probabilities, dispatching between JAX and NumPy implementations.
+    """
+    if JAX_AVAILABLE:
+        is_probs_jax = isinstance(probs, jax.Array)
```
Comment (mudit2812) [inline pennylane/devices/qubit/sampling.py:502]: `jax.numpy.array` does not return a `jax.Array`, but rather a `jaxlib.xla_extension.ArrayImpl`. Instead, we can use `qml.math.get_interface` here:
```suggestion
        is_probs_jax = qml.math.get_interface(probs) == "jax"

**Review by mudit2812 — CHANGES_REQUESTED** [review]
Looks good so far 🚀 . I've left a few comments that I think should be addressed before I can approve.

**Review by JerryChen97 — COMMENTED** [review]
Seems all resolved but somehow Github doesn't auto reset

**Comment by JerryChen97** [comment]
Hi @lillian542 my implementation involved some refactoring over your previous design as well as some tests; fortunately all the tests seem smooth for now but I just want to double check to make sure there's no explicit missing point on my side

File: tests/devices/qubit/test_sampling.py [hunk 5]
```diff
@@ -84,42 +84,42 @@ def test_sample_state_basic(self, interface):
 
     @pytest.mark.jax
     def test_prng_key_as_seed_uses_sample_state_jax(self, mocker):
-        """Tests that sample_state calls _sample_state_jax if the seed is a JAX PRNG key"""
+        """Tests that sample_state calls sample_state if the seed is a JAX PRNG key"""
```
Comment (mudit2812) [inline tests/devices/qubit/test_sampling.py:87]:
```suggestion
        """Tests that sample_state calls _sample_probs_jax if the seed is a JAX PRNG key"""
```

File: tests/devices/qubit/test_sampling.py [hunk 6]
```diff
@@ -84,42 +84,42 @@ def test_sample_state_basic(self, interface):
...
         import jax
 
         jax.config.update("jax_enable_x64", True)
 
-        spy = mocker.spy(qml.devices.qubit.sampling, "_sample_state_jax")
+        spy = mocker.spy(qml.devices.qubit.sampling, "_sample_probs_jax")
         state = qml.math.array(two_qubit_state, like="jax")
 
-        # prng_key specified, should call _sample_state_jax
+        # prng_key specified, should call sample_state
```
Comment (mudit2812) [inline tests/devices/qubit/test_sampling.py:95]:
```suggestion
        # prng_key specified, should call _sample_probs_jax
```

File: pennylane/devices/qubit/sampling.py [hunk 7]
```diff
@@ -471,67 +471,58 @@ def sample_state(
...
-    state,
-    shots: int,
-    prng_key,
-    is_state_batched: bool = False,
-    wires=None,
-    seed=None,
-) -> np.ndarray:
+def _sample_probs_jax(probs, shots, num_wires, is_state_batched, prng_key=None, seed=None):
     """
     Returns a series of samples of a state for the JAX interface based on the PRNG.
```
[the remark this answers is not in the document]
Comment (JerryChen97) [reply pennylane/devices/qubit/sampling.py:527]: Good catch!!! Thanks a lot! This happend due to lots of jump between before

File: pennylane/devices/qubit/sampling.py [hunk 8]
```diff
@@ -471,78 +471,88 @@ def sample_state(
...
 
     Args:
         state (array[complex]): A state vector to be sampled
```
Comment (mudit2812) [inline pennylane/devices/qubit/sampling.py:550]:
```suggestion
        probs (array): The probabilities to sample from
```

File: pennylane/devices/qubit/sampling.py [hunk 9]
```diff
@@ -471,78 +471,88 @@ def sample_state(
...
         shots (int): The number of samples to take
-        prng_key (jax.random.PRNGKey): A``jax.random.PRNGKey``. This is
-            the key to the JAX pseudo random number generator.
+        num_wires (int): The number of wires to sample
         is_state_batched (bool): whether the state is batched or not
-        wires (Sequence[int]): The wires to sample
-        seed (numpy.random.Generator): seed to use to generate a key if a ``prng_key`` is not present. ``None`` by default.
+        prng_key (Optional[jax.random.PRNGKey]): An optional ``jax.random.PRNGKey``. This is
+        seed (Optional[int]): A seed for the random number generator.
```
Comment (mudit2812) [inline pennylane/devices/qubit/sampling.py:555]:
```suggestion
        prng_key (Optional[jax.random.PRNGKey]): An optional ``jax.random.PRNGKey``. This is
            the key to the JAX pseudo random number generator. Only for simulation using JAX.
        seed (Optional[int]): A seed for the random number generator. This is only used if ``prng_key``
            is not provided.
```

**Review by mudit2812 — APPROVED** [review]

**Comment by mudit2812** [comment]
First PR almost in 🎉 !

**Review by lillian542 — APPROVED** [review]
🚀
